web-animations-2/Overview.bs

```diff
@@ -143,10 +143,51 @@ Along with the following updated description:
 
 </div>
 
+<h3 id="timelines">Timelines</h3>
+
+Add:
+
+> The <dfn lt="timeline duration">duration</a> of a timeline gives the
+> maximum value a timeline may generate for
+> [=timeline current time|current time=]. This is used to calculate the
+> [=iteration duration=] when the string <code>auto</code> is specified such that
+> the animation fills the available time.
```

Comment:
A few thoughts here:

* Firstly, this is part of the model description, as opposed to the API, so we generally try to phrase things in terms of the concepts rather than the specific interface types. (I'm not sure if that's a good idea after all, but that's the state of affairs at the moment.)
* Secondly, I think we need to say where the `auto` value is specified -- in this case I think we are talking about the target effect of an animation association with the timeline. However...
* Thirdly, in level 2 of the spec, `auto` is also used to allow groups to have a duration calculated based on the lengths of their children. At the model level, `auto` means "use the intrinsic iteration duration".

So, I think this sentence should probably be something like:

> This is used in calculating the intrinsic iteration duration of the target effect of animations associated with the timeline.

Which brings us to something I think you might not want to hear: I think we probably need to specify how these different cases interact.

I wonder if can do something like the following:

1. Redefine the _intrinsic iteration duration_ as being either a number or a percentage/proportion or possibly a calc expression of both.
2. Define the _intrinsic iteration duration_ of keyframe effects (or all animation effects that don't supply their own definition for intrinsic iteration duration) as being 100% (instead of 0s).
3. Rework the definition of the _intrinsic iteration duration_ of group effects to similarly produce either a number or proportion or potentially a `calc()` if we allow that.

    In a sense, these steps produce a bottom-up initial calculation of the target effect's intrinsic iteration duration.
4. (Optionally define the timeline duration of a document timeline as being zero instead of null since I think that might make the spec simpler if there's no particular need for `null`.)
5. Define a "resolve computed intervals" procedure where we start with the target effect of an animation, and resolve its duration, start delay, and end delay based on the particular combination of its:

    * Timeline duration
    * Specified iteration duration or intrinsic iteration duration
    * Specified start delay and end delay

    This would be similar to the conversion procedure defined in this spec. In the simple case, nothing would be need to be resolved as everything would already be in milliseconds. Similarly, when we have everything in percentages, there would be nothing to do.

    In other cases, where we have a millisecond duration timeline (e.g. `DocumentTimeline` with its 0ms duration) and proportion times, we would resolve the proportions to milliseconds. (This would allow a `KeyframeEffect` with an `auto` duration to continue producing a resolved duration of 0s.)

    In yet other cases where we have a percentage timeline and millisecond times, we would convert the milliseconds to percentages as we have in the conversion algorithm in this spec.

    Then of course, there are the odd cases where we have a mix we don't support and where we need to use some sensible behavior. I think it's worth at least thinking about how we could support `calc(50% + 2s)`.
6. Recursively apply this behavior to the target effect's child effects.

I'm really sorry. I know you wanted to keep this patch as simple as possible just to get this working for scroll timelines, but I think we need to work out how this integrates with group effects.

On the positive side, if the above approach works, we'll have solved percentage based timing in general and made something really useful, I think.

I think it would also extend naturally to other use cases like `KeyframeEffect`s where the keyframe offsets are specified in milliseconds. We would simply plug those values into the calculation of the intrinsic iteration duration of the keyframe effect and the rest would just work.

Reply:
I should clarify step 6:

> 6. Recursively apply this behavior to the target effect's child effects.

I don't think we'd need to actually write an algorithm to recursively visit the children. Instead we could just write a definition for _computed iteration duration_, _computed start delay_ etc. in terms of their parent time source's intrinsic iteration duration and then we'd naturally get the recursive behavior (i.e. the bottom-up followed by top-down traversal).

Reply:
Hi Brian,
I've jumped in to assist Rob on the spec work and have a new pull request (https://github.com/w3c/csswg-drafts/pull/6337), which hopefully addresses the outstanding issues.

Some of the key changes:
* intrinsic iteration duration updated to be either a percent or time.  In the case of a progress based timeline, the percent is simply 100% / iteration_count.  Note that in the future, if start and end delay are updated to allow percentages, the intrinsic iteration duration becomes (100% - start - end) / iteration_count.  Once start and end percentages are supported, we will need to be able to handle start + end > 100%, which would otherwise result in a negative intrinsic iteration duration.  This should be fairly easy to handle. Just raising the issue now so that we don't lose track of it.
* Updated to procedure for calculating the intrinsic iteration duration to support group/sequence effects as well as progress based timelines.

Note that mixing times and percentages opens up some awkward edge cases that are not easily addressed (examples below).  Ultimately, we can easily convert form entirely time based to percentage based.  The inverse is only possible if knowing the effect end time.  For now, I was planning to stick with Rob's suggestion of start and end being set to 0 if the timeline is progress based.

Currently the iteration duration can be a time or 'auto'.  In the future, we could extend to time 'auto' or percent. I would prefer to defer introducing percentages outside of 'auto' to avoid complications that would be introduced until all of the details can be worked out.  The value for effect end time depends on iteration duration, which in turn could be the intrinsic iteration duration if 'auto'.  The intrinsic iteration duration in turn can depend on effect end time. 

Let's start with a few cases that would work, and them move to the ugly cases:

Case 1:

.anim {
  animation: parallax 1s 0.25s scroll-timeline;
}
.anim2 {
  Animation: parallax 2s scroll-timeline; 
}

Here everything is time based, so the conversion to percentages for use with a progress based scroll timeline is trivial.

anim: iteration duration = 100% * (1s / 1.5s) = 66.67%, start delay = end delay = 100% * 0.25s / 1.5s = 16.67%
anim2: iteration duration = 100%, start delay = end delay = 0%

Case 2:

.anim {
  animation-name: parallax;
  animation-delay: 20%; /* once supported as a percent */
  Animation-timeline: scroll-timeline;
}

intrinsic iteration duration = 100% - 20% = 80% = iteration duration

Case 3:

.anim {
  animation-name: parallax;
  animation-duration: 1s;
  animation-delay: 20%;
}

Now we would need to be prepared to a bit more work to resolve the timing by collecting times and percentages.

end time = 1s + 20% = 100% ==> 1s  = 80% ==> 1.2.s = 100%.

Note: at this stage we can get inconsistent timings (e.g. 0s animation + delay != 100%). 

Case 4: 

.anim {
  animation-name: parallax;
  animation-duration: 75%;
  animation-delay: 1s;
}

Similar approach to case 3.

end time = 75% + 1s = 100% ==> 1s = 25% ==> 4s = 100%.

Case 5:

.anim {
  animation-name: parallax;
  animation-delay: 2s;
  Animation-iteration-count: 2;
}

Now for an ugly case.

end time = 2*(intrinsic iteration duration) + 2s  = 100% 
==> 2*((100% - 2s))/2 + 2s = 100%
==> 100% = 100%

Here the intrinsic iteration duration is arbitrary. It may make sense to use the minimum solution (0s/0%) in this case. I welcome further discussion in this area, but would prefer to defer allowing mixing of percentages and times to a followup pull request.  I believe have a pretty solid base, which can be extended in a way that won't require backpedaling.

Reply:
Hi Kevin,

Thank you so much for following up on this. Unfortunately I'm going to be quite busy for the next few days. Do you mind if I get back to you on this after that? Please feel to ping me if I'm taking too long or if you have any particular time constraints.

Reply:
No rush.

Reply:
Thanks for your patience on this. I've finally reviewed #6337 and it looks good.

Responding to your comment,

> Case 1:
> 
> .anim {
> animation: parallax 1s 0.25s scroll-timeline;
> }
> .anim2 {
> Animation: parallax 2s scroll-timeline;
> }
> 
> Here everything is time based, so the conversion to percentages for use with a progress based scroll timeline is trivial.
> 
> anim: iteration duration = 100% * (1s / 1.5s) = 66.67%, start delay = end delay = 100% * 0.25s / 1.5s = 16.67%
> anim2: iteration duration = 100%, start delay = end delay = 0%

I'm curious why start and end delay end up being made the same. Shouldn't the `0.25s` from the `animation` declaration be mapped to start delay only?

Overall though it's looking great. I'm a bit uncomfortable about having combinations of content that don't work now but which we expect to work in future.

It would be good to add another PR at some point filling out those cases. I'd even be to say, for example, when an animation is directly or indirectly associated with a progress-based timeline, if it uses a combination of percentages and times, we just ignore the times (i.e. use their default values) if need be. At least that's consistent, I guess.

Maybe that's roughly equivalent to what you have, though. From what I tell though, with #6337 if you specified duration 5s, start delay 5%, end delay 2s, we'd end up ignoring _both_ start delay and end delay as opposed to simply ignoring the time-based end delay.

Reply:
Oops...yes case 1 should be  a 1.25s animation which then maps to 100% so that the iteration duration and start delay can be converted to percentages.